aziot/src/init.rs

```diff
@@ -0,0 +1,948 @@
+		line.push_str(&password);
+		Ok(password.len())
+	}
+}
+
+fn choose<'a, TChoice>(
+	stdin: &mut impl Reader,
+	question: &str,
+	choices: &'a [TChoice],
+) -> Result<&'a TChoice, crate::Error> where TChoice: std::fmt::Display {
+	println!("{}", question);
+
+	#[allow(
+		clippy::cast_possible_truncation,
+		clippy::cast_precision_loss,
+		clippy::cast_sign_loss,
+	)]
+	let max_choice_width = (choices.len() as f64).log10() as usize + 1;
+
+	println!();
+	for (i, choice) in choices.iter().enumerate() {
+		println!("{:>max_choice_width$}: {}", i + 1, choice, max_choice_width = max_choice_width);
+	}
+	println!();
+
+	let prompt_question = format!("Enter a choice [1-{}]", choices.len());
+
+	loop {
+		let answer = prompt(stdin, &prompt_question)?;
+		let answer =
+			answer.parse().ok()
+			.and_then(|choice: usize| choice.checked_sub(1))
+			.and_then(|choice| choices.get(choice));
+		if let Some(answer) = answer {
+			return Ok(answer);
+		}
+
+		println!("Invalid choice.");
+	}
+}
+
+fn prompt(stdin: &mut impl Reader, question: &str) -> Result<String, crate::Error> {
+	println!("{}", question);
+
+	let mut line = String::new();
+
+	loop {
+		stdin.read_line("> ", &mut line)?;
+		if line.is_empty() {
+			continue;
+		}
+
+		println!();
+
+		return Ok(line);
+	}
+}
+
+fn prompt_secret(stdin: &mut impl Reader, question: &str) -> Result<String, crate::Error> {
+	println!("{}", question);
+
+	let mut line = String::new();
+
+	loop {
+		stdin.read_secret("(echo disabled) > ", &mut line)?;
+		if line.is_empty() {
+			continue;
+		}
+
+		println!();
+
+		return Ok(line);
+	}
+}
+
+fn parse_manual_connection_string(connection_string: &str) -> Result<(String, String, Vec<u8>), String> {
+	const HOSTNAME_KEY: &str = "HostName";
+	const DEVICEID_KEY: &str = "DeviceId";
+	const SHAREDACCESSKEY_KEY: &str = "SharedAccessKey";
+
+	let mut iothub_hostname = None;
+	let mut device_id = None;
+	let mut symmetric_key = None;
+
+	for sections in connection_string.split(';') {
+		let mut parts = sections.split('=');
+		match parts.next().expect("str::split always returns at least one part") {
+			HOSTNAME_KEY => iothub_hostname = parts.next(),
+			DEVICEID_KEY => device_id = parts.next(),
+			SHAREDACCESSKEY_KEY => symmetric_key = parts.next(),
+			_ => (), // Ignore extraneous component in the connection string
+		}
+	}
+
+	let iothub_hostname = iothub_hostname.ok_or(r#"required parameter "HostName" is missing"#)?;
+
+	let device_id = device_id.ok_or(r#"required parameter "DeviceId" is missing"#)?;
+
+	let symmetric_key = symmetric_key.ok_or(r#"required parameter "SharedAccessKey" is missing"#)?;
+	let symmetric_key =
+		base64::decode(symmetric_key)
+		.map_err(|err| format!(r#"connection string's "SharedAccessKey" parameter could not be decoded from base64: {}"#, err))?;
+
+	Ok((iothub_hostname.to_owned(), device_id.to_owned(), symmetric_key))
+}
+
+fn convert_path_or_uri_to_uri(value: String) -> Option<String> {
```

Comment:
super minor, but there's technically no reason why a user can't use a `pkcs11:` prefixed string when they're not supposed to. I know there's no reason why someone _would_ do that, but maybe it makes sense to add a `allow_pkcs11: bool` to the function signature.

Reply:
Fixed. I made it into two functions that properly parse URLs of the target type (`aziot_keys::PreloadedKeyLocation` for keyd and `url::Url` for certd).